### ISSUE TITLE
Determine what ambient telemetry will look like

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#42320]
In sidecars, we have `reporter=source|destination`. In ambient, things are a bit less clear.
We could have up to 4 proxies along the path, and 2 (ztunnel) may not be doing L4 processing at all.
However, the ztunnels can provide L4 telemetry and could even provide L4 traces to some extent (since we can propogate tracing headers via HBONE).

This issue tracks designing what telemetry should be emitted in ambient

size: L, type: design

related WI: https://github.com/istio/istio/issues/45676

**Comment by kyessenov** [comment]
For ztunnel, there's a gap with Telemetry API because there's no CEL in Rust so customizations are not available (cc @zirain ).

**Comment by howardjohn** [comment]
I think the limitations may be beyond Rust vs Envoy if we include the full scope of Telemetry customization. We probably don't want per-workload overrides on node proxy

**Comment by kyessenov** [comment]
Yes, I'm not suggesting that, but I think it's fairly straightforward to maintain Waypoint full support for telemetry API as server-side for the producer and client-side for the consumer.

For node proxies, I am not convinced we need to present L4 Istio metrics. The main audience would be the node operators since they are multi-tenant and perhaps not even visible to the application developers. The focus there should be on the infrastructure metrics, like the Envoy default metric set.

**Comment by SkyfireFrancisZ** [comment]
Lin and Ethan will find the right person to tackle this workitem.

**Comment by krisztianfekete** [comment]
> ambient should take this opportunity to change the bootstrap histogram bucketing.
       > Best would be to not use Envoy defaults (Prom defaults would likely be better) and allow custom bucketing, if possible.

Custom bucketing is not (yet) possible, but will be available with the new native histogram metric type. This is available since Prometheus v2.40 as an experimental feature. I will check the design docs regarding compatibility questions.
Until that, fixed buckets is fine, with reasonable defaults (based on extensive testing).

> Perhaps the L4 telemetry could add a label indicating whether it is strictly L4 or carries L7 info. That would make it easy to query for a complete L4 view, while also making it easy eliminate L4 metrics that would presumably be reported as L7 requests as well.

There's already a `request_protocol` label, but a new one dedicated to distinguish between L4 and L7 can make sense, and would not be an issue of cardinality. This way, users would be able to run the same queries regardless of the layer  and would be switch between the two "modes".

> I think source and/or destination reporting should be the goal, no more than two reporters reflecting the source and destination "view" of a request.
      > Is this something that becomes less important with correlation IDs and baggage? That's sort of hard to 
         enforce/track/make assumptions about on the Istio data plane side, globally.

Here we are just talking about metrics, and that should be handled separately as that's one of the easiest, cheapest observability signal to get, and most users stop at this point (even if it's only temporarily), so they should be useful by themselves too. 

> collection and storage does impact users and many end up dropping metrics or turning off collection entirely

Only a reasonable set of metrics should be enabled by default. Regular Istio did not get this totally right, as [these can scare](https://istio.io/latest/docs/ops/best-practices/observability/) people off (or they can ignore these and suffer the consequences).

This is another example that I feel like should be a default as only a very small subset of users would like to have this: https://docs.solo.io/gloo-mesh-enterprise/latest/observability/tools/prometheus/production-setup/#remove-cardinality-labels

> this is where I think having a configurable collector that is separate from Istio makes a lot of sense. Istio emits what it knows, there's a collector that ingests the emitted metrics, and you can drop or exclude what you like there before the metrics are forwarded to your store, based on your needs.

OTel is just an implementation detail, and while it's getting more and more popular, most users are not using it yet but relying on Prometheus as that's still the de-facto standard. I think in terms of UX, we should focus on getting this right with Prometheus first, then OTel. It's quite easy as Prometheus Receiver is quite close to Prometheus.

**Comment by jshaughn** [comment]
@howardjohn, As a follow-up to the working group meeting, if you and Louis have information to share please add it here in some way, thanks!

**Comment by keithmattix** [comment]
Not stale

**Comment by linsun** [comment]
@justinpettit can you identify some resource for this?  Would love to have a high level design for telemetry before ambient announces some sort of beta.

**Comment by jshaughn** [comment]
Thinking more on Ambient metrics I want to follow up on my initial comments.   I do think the existing, core Istio metrics should be maintained for the foreseeable future.  Doing this should be more seamless as people migrate to ambient, and also let OTel work evolve.  At the same time, for ambient namespaces maybe we can treat telemetry a little differently, and mostly leave sidecar proxy telemetry alone.  It seems to me that ambient users will move quickly to ambient, and mainly leave sidecars behind.  And from a telemetry perspective I think we can probably ignore telemetry from sidecars within an ambient namespace (will there even be sidecars in an ambient namespace?).

**Eliminate Source/Dest Reporters**
The source and destination reporting of sidecar proxies is rich, but wasteful.  For successful requests the source and dest reporting is mostly duplicated.  That results in double the processing, double the scraping, double the storage, and sometimes the need to de-dupe after querying.  The Ambient arch is different, there are no source and dest proxies and reporting should not try and emulate it.  A new "reporter" has been suggested in the past, and I now think that `reporter = "ambient"`makes sense, because it's accurate.  It also makes it easy to query ambient telemetry via a single label-value pair.  The reporter could also be set to "ztunnel" or "waypoint", but that may be better as something like `reporter_app` (Note, I think currently tcp metrics have `app="ztunnel"`, I'd recommend that change to what's proposed above).

I believe today ambient is always using `reporter="destination"`.  This may seem like a good idea for some amount of back-compat, but I think it's misleading.  For example, it's quite possible to have `reporter="destination"` yet no destination_workload, because in reality the request never made it to a dest workload.

If there is a sidecar proxy in an ambient namespace the telemetry could continue to report as it does today, independently of the ambient telemetry. Or, maybe it could be disabled completely, as some telem consumers, like Kiali, may elect to ignore it.

Removal of source and dest reporting will result in significant savings in telemetry capture, storage and processing.  The main concern is whether Ambient telem is able to capture the union of what the source and dest reporting did with sidecar proxies.   For example, sidecar client reporting captured requests that never made it to a destination proxy, and destination reporting provided security policy.

**New bucketing for Histograms**
A significant number of users disable Istio's histograms because they are too expensive.  Currently, Istio bootstraps using Envoy's [default histogram bucketing](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto), with 19 buckets:

    [ 0.5, 1, 5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000, 30000, 60000, 300000, 600000, 1800000, 3600000 ]

Prometheus defaults cut the number almost in half:

    [ .005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10 ]

[OTel HTTP defaults](https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#metric-httpserverrequestduration) are similar to Prom, but with 14 buckets:

    [ 0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10 ].

Each unique time-series gets one more dimension for each bucket (+ 1 more for "inf"), so in a simple case, say requests A->B, there are currently 20 bucket time-series generated, with each updated for a value <= the bucket value.  For example, using request_duration_milliseconds, for an 1100ms entry, the first 10 buckets are updated (+ 1 more).  And note, that is for *each* unique set of attributes.

In my opinion Envoy has too many buckets, bloating the storage and slowing down updates, queries and calculations. Prom and OTel are better but currently set for seconds, whereas  Istio is currently reporting in milliseconds.

It's not totally clear what to do here. Also, Prometheus has just introduced experimental support for "Native Histograms", which is different from what I described above, and may help out substantially.  It is something that should be tracked for possible use with ambient GA.  But for now, I'd suggest that when installing Istio with the ambient profile, we adhere to the OTel bucketing, adjusted for milliseconds.  This will still reduce the overall number of time-series while adding some standardization.  Id ignore support for custom bucketing and hope that native histogram support eventually solves the histogram problem.

**Removals**
This is biased towards what we currently need in Kiali, but I think some attributes could be removed from the standard metrics.  This may not help much with cardinality, but it would help with storage, and would simplify the metrics.  For Kiali users looking to save metric storage space, we recommend the following labeldrops:

chart|destination_app|destination_version|heritage|.*operator.*|istio.*|release|security_istio_io_.*|service_istio_io_.*|sidecar_istio_io_inject|source_app|source_version

Minimally, I think the "canonical" attributes remove the need for `*_app` and `*_version`.

Note that the nature of waypoints should already reduce cardinality significantly, as the `pod`attribute should have far fewer values, especially for instances where app pods saw a lot of churn.

There are also a lot of metrics that could likely be turned off by default.  In Kiali the following metrics that are currently unused:

istio_agent_.*|istiod_.*|istio_build|citadel_.*|galley_.*|pilot_[^p].*|envoy_cluster_[^u].*|envoy_cluster_update.*|envoy_listener_[^dh].*|envoy_server_[^mu].*|envoy_wasm_.*

To summarize, the above doesn't recommend any sort of fundamental shift in telemetry for Ambient, but I think would let Ambient report in a more native way, address a legacy issue in histogram reporting, reduce the telemetry footprint, and require only modest changes for consumers.

**Comment by linsun** [comment]
Hi @jshaughn thank you so much for your analysis! Would it be possible for you to discuss your recommendation in an upstream ambient meeting?  I'm interested in reaching some consensus for beta for L4 and then L7 later. Wasn't clear to me which of your 3 major bullets apply to L4 also? Maybe only "Eliminate Source/Dest Reporters"?

@srampal  or @justinpettit do you have any update on this issue?

**Comment by jshaughn** [comment]
> Overwriting the buckets is possible in multiple ways already, so I am not sure if it worth changing them this way. It's also a bit unrelated to Ambient itself.

Can I ask how this is done?  In my experience Istio always bootstraps back to the Envoy defaults.  If there is an easy way to customize the bucketing for istio_request_duration_milliseconds (and other histograms), then problem solved.

As for dropping metrics, I understand that there are lots of compat issues and in general this should just be made easily configurable.  My main theme is that many users experience overwhelming amounts of telem, and as such it becomes unusable.  As Istio pivots towards Ambient, it will be good to acknowledge this fact and work on trimming, as is possible.

In what I wrote above, with respect to what can be applied strictly to Ambient, the focus would be on maintaining existing metrics and the `reporter` attribute.

**Comment by krisztianfekete** [comment]
@jshaughn, I've had luck with reconfiguring it via the `sidecar.istio.io/statsHistogramBuckets` annotation previously.

**Comment by linsun** [comment]
Pinged @lei-tang @jshaughn on slack - can you give some update on ambient telemetry for beta in this week or an upcoming ambient meeting?

**Comment by lei-tang** [comment]
As (source, destination) pairs can have large cadinality, from the memory consumption and transmission costs perspective, it makes sense to remove them from metric labels. For example, if source and destination each has a cadinality of 20, having source and destination labels will lead to 20x20=400 times more memory consumption and transmission costs.

A major advantage of Ambient is its efficiency. Removing some high cardinality labels and some non-essential metrics will greatly improve the efficiency - less memory consumption, less metric transmission costs, and less memory-related crashes.

**Comment by jshaughn** [comment]
It seems that @lei-tang and I agree that in Ambient it doesn’t make sense to try and emulate source and destination reporting.  The idea of reporter=”source” and reporter=”destination” is tied to having source and destination proxies, each processing a request.  Although, I’m not sure the issue is as much the source X dest cardinality, we still need to know the source and destination information for the requests, but rather that with two reporters for the same request, the cardinality doubles (source X dest X 2).  

Let me take a shot at a succinct set of metric telemetry suggestions for Ambient beta.

- Istio implementations of ztunnel and waypoints should maintain the current Istio metrics, with minor modifications, as described in following points.
  - This ignores my suggestions above, for slimming existing metrics by dropping fields and/or changing the bucketing for histograms.  While useful, that’s actually outside of Ambient proper.
  - This does not prevent additional, next-gen metrics being introduced as a result of other work, most likely based on emerging OTel standards for HTTP request reporting.  When ready, a crossover period could emerge where old and new metrics could both be generated (or easily disabled).  Non-Istio implementations may want to offer their own metrics, or focus solely on OTel, etc, but for continuity Istio impls should continue to be generated for traffic handled by Ambient telemetry.
- Ztunnel and Waypoints should not use reporter=”source” or reporter=”destination”, as the Ambient components are neither a source or destination sidecar proxy (although, to some degree I guess it could be thought of that way if a request originates or terminates in a proxy, outside of the Ambient namespace). There are a couple of options but I think this one is probably sufficient:
  - Have ztunnel use reporter=”ztunnel” and waypoints use reporter=”waypoint”.
    - This should be sufficient if the different components report to distinct metrics.  For example, ztunnel writes just to TCP metrics and waypoints write only to request metrics.  Otherwise it may be best to have both use reporter=”ambient” and add a new label like reporter_component=”ztunnel | waypoint”.
    - This allows consumers to easily identify ambient telem in a query using a single label-value pair, and distinguish it from sidecar-proxy telem.
  - This idea is predicated on the idea that telemetry can be reliably generated one time for one request.  This already seems to be working for the L4 level.  The real question is whether it can be done by the waypoint?  I think it would mean that for a single request the waypoint would write telemetry only once, at one of these times:
    - Immediately for requests that failed to establish a destination workload
    - When an upstream connection is lost
    - After a request is received and all information is collected
- Currently, I think all ambient telemetry is setting reporter=”destination”.  I think this is misleading and may cause confusion for legacy consumers.  But, if desired it could remain as long as there were another label identifying that ztunnel or waypoint were doing the reporting.

That's it, given the technical feasibility, especially for waypoints, I think for Ambient beta we could just be talking about existing telem with a little bit of label shuffling.  The nature of Ambient architecture would already cut telem reporting in 1/2.  The main thing to remember is that users need the telemetry to be as robust as possible, especially in failure scenarios and latency reporting.

@linsun It would be great if you could get some of the ztunnel and/or waypoint contributors to comment...

**Comment by lei-tang** [comment]
Adding a new label "reporter_component" to OSS ambient adds to its complexity and is not consistent with OSS Istio, which does not have this new label. 

I prefer for ztunnel, reporter=”ztunnel”; for waypoint proxy, reporter=”waypoint”, because it is simple and clear.

**Comment by bleggett** [comment]
> Adding a new label "reporter_component" to OSS ambient adds to its complexity and is not consistent with OSS Istio, which does not have this new label.

That's not necessarily a problem, OSS ambient looks nothing like OSS sidecar in practical terms anyway.

My concern is if we (eventually) end up with something like

- 1 ztunnel (ztunnel A) per node
- N waypoints per node

and ztunnel A is proxying L4 traffic to all ambient pods on the node _as well as_ terminating HBONE traffic bound to those N waypoints - it gets potentially a bit trickier to collate the info.

Thinking about it, it's probably it's moot either way tho, as we will eventually _have_ to append more specific context than `reporter=waypoint|ztunnel` for ambient.

My only concern is that when we are forced to do that, we will make it hard for ourselves to support parallel telemetry views for sidecar and ambient traffic in the same cluster.

**Comment by krisztianfekete** [comment]
> I prefer for ztunnel, reporter=”ztunnel”; for waypoint proxy, reporter=”waypoint”, because it is simple and clear.
+1

**Comment by jshaughn** [comment]
Today, with sidecar injection, for a single request from A to B there are 2 time-series generated for each relevant metric (requests_total, tcp_sent_bytes, tcp_received bytes, etc), sidecar-A with `reporter="source"` and sidecar-B with `reporter="destination`.  Is there consensus that with Ambient, for a single request from A to B, that we would expect 1 time-series generated for each relevant metric?  Potentially with `reporter="ztunnel"` (for L4 metrics, like tcp_sent_bytes) or `reporter="waypoint"` (for L7 metrics, like request_total).

**Comment by mikemorris** [comment]
From a cursory glance (caveat that I haven't fully read all the comments) I'd expect that `reporter="source"` and `reporter="destination"` should generally retain similar usage and not be overloaded with a _less_ descriptive new term for Ambient - `reporter="ztunnel"` fails to describe whether the ztunnel is reporting from the source or destination, and the actually becomes _more_ important for ambient data flows due to the possibility of waypoint hops.

I'd be supportive of adding something like a `reporter_component` label and using `reporter_component="sidecar"` for consistency with the traditional model to address @lei-tang's concern. With this, a typical flow (using a "sandwich" deployment) may look like:

| client zt | waypoint zt -> waypoint | waypoint -> waypoint zt | destination zt |
| --- | --- | --- | --- |
| reporter="source" | reporter="destination" | reporter="source" | reporter="destination" |
| reporter_component="ztunnel" | reporter_component="ztunnel" | reporter_component="ztunnel" | reporter_component="ztunnel" |

I'd expect the same-node waypoint <-> ztunnel hops to be negligible, but the biggest gap here is the processing time within the waypoint itself. For this reason, I _think_ there should ideally be a "parent" tracing span capturing the entire client zt -> destination zt timing, although that's likely only possible to emit from the destination zt, and only if the waypoint forwards timing information from the first leg.

| client sidecar | destination sidecar |
| --- | --- |
| reporter="source" | reporter="destination" |
| reporter_component="sidecar" | reporter_component="sidecar" |

**Comment by kyessenov** [comment]
I don't think we need `reporter_component` - all monitoring systems allow tagging metrics with the component the metrics  come from. This is usually referred to as a "monitored resource".

The main value of the two `reporter`s is that it provides an easy way to compare and the difference is meaningful since it reflects the network impact. There's not much value in comparing ambient telemetry to sidecar telemetry, so I think you can pick a completely different set of labels.

**Comment by bleggett** [comment]
> From a cursory glance (caveat that I haven't fully read all the comments) I'd expect that reporter="source" and reporter="destination" should generally retain similar usage and not be overloaded with a less descriptive new term for Ambient - reporter="ztunnel" fails to describe whether the ztunnel is reporting from the source or destination, and this actually becomes more important for ambient data flows due to the possibility of waypoint hops and therefore multiple source and destination points within a single flow.

Yes +1 on this, this is my main concern. Component != direction or origin, as long as we aren't conflating those things, we should be good.

**Comment by howardjohn** [comment]
Chatted with @louiscryan a bit, here is where we arrived. Primary focus is retaining metrics value while keeping cost reasonable.

I setup a small spreadsheet to put in theoretical numbers and analyze different approaches. Here are the primary approaches I considered:

![image](https://github.com/istio/istio/assets/623453/df79eef0-fc4e-437d-8668-31e37854b27a)

Here the boxes are the actual proxy reporting the metric, and the source.

Note: the ones with empty spot are derived from others.

For some reason I cannot upload to the team drive with some permission issue, so attached here
[Ambient telemetry analysis.xlsx](https://github.com/istio/istio/files/14485990/Ambient.telemetry.analysis.xlsx)

Summarized (waypoint double == top option, waypoint single == bottom in the picture)

| Metric types                  | Reporting Pods | Reporters | Metric count | Destinations | Sources | Series    | Cost    |
| ----------------------------- | -------------- | --------- | ------------ | ------------ | ------- | --------- | ------- |
| Sidecar HTTP                  | 1000           | 2         | 64           | 7            | 1       | 896,000   | $7,168  |
| Sidecar TCP                   | 1000           | 2         | 4            | 7            | 1       | 56,000    | $448    |
| Ztunnel                       | 25             | 2         | 4            | 100          | 25      | 500,000   | $4,000  |
| Waypoint double (client part) | 30             | 1         | 64           | 1            | 70      | 134,400   | $1,075  |
| Waypoint double (server part) | 30             | 1         | 64           | 10           | 1       | 19,200    | $154    |
| Waypoint double               |                |           |              |              |         | 153,600   | $1,229  |
| Waypoint single               | 30             | 1         | 64           | 10           | 70      | 1,344,000 | $10,752 |
| Waypoint single + ztunnel     |                |           |              |              |         | 1,844,000 | $14,752 |
| Waypoint double + ztunnel     |                |           |              |              |         | 653,600   | $5,229  |

Broad notes:
* Histograms are killer. Native histogram support, or even dropping buckets, may yield huge wins -- but they scale linearly so its not foundational
* Ztunnel cost is pretty small relative to waypoints
* An 'off the shelf' waypoint implementation outside of Istio would likely report only the right side of the 'waypoint double'
* Much of the cost comes from reporting pod_name in the label. This cannot be dropped out of the box in simple prometheus setup, but it is possible with prometheus https://karlstoney.com/federated-prometheus-to-reduce-metric-cardinality/ and generally any metric backend.

There is discussion about consolidating {workload, canonical service, app} and {version, canonical revision}; since these are generally the same values, these do not meaningfully impact cardinality which is the primary driver for costs in metrics. Given this, its probably best to keep all the labels for compatibility.

**Comment by kyessenov** [comment]
"Waypoint-double" reporting is qualitatively different since you cannot correlate the two metrics easily. Intuitively, for full fidelity with sidecars, you need a series per every network hop so it's always O(hops). You can sacrifice some fidelity via an intermediary LB service (so instead of M*N tuples you have M+N tuples), but it is not equivalent.

**Comment by lei-tang** [comment]
I prefer the "waypoint single" approach over the "waypoint double" approach because: 

- The "waypoint single" approach provides the source -> destination information. The "waypoint double" approach does not provide the source -> destination information. 

Take the following data points in the "waypoint double" approach as an example: source 1 -> waypoint, waypoint -> destination 2, source 2 -> waypoint, waypoint -> destination 1. Such data points do not provide the source -> destination information: e.g., from these data points, a user can not tell the actual source and destination is source 1 -> destination 1 or source 1 -> destination 2. 

- The "waypoint single" approach is simpler than the "waypoint double" approach.

**Comment by mikemorris** [comment]
Would it be worthwhile to consider some labeling scheme and configuration such that it could be _possible_ to enable "waypoint double" telemetry even if "waypoint single" is default and recommended for cost control in production?

**Comment by kyessenov** [comment]
@mikemorris is there a material difference between the "double" telemetry and a "single" telemetry with cardinality reduction techniques applied (e.g. erasing destination labels, grouping of response codes, etc)?

**Comment by mikemorris** [comment]
> @mikemorris is there a material difference between the "double" telemetry and a "single" telemetry with cardinality reduction techniques applied (e.g. erasing destination labels, grouping of response codes, etc)?

I think I don't have a deep enough understanding to answer that well tbh.

> if the waypoint is having issues we are likely to use a more operation metric (CPU, queue lengths, etc) to diagnose issues

My concern is mostly around whether ^ is sufficient (and an expected triage workflow) for Istio operators in this situation.

**Comment by jshaughn** [comment]
I prefer waypoint-single as I don't think that in general users want to see the infrastructure in the telemetry.  They want to see that traffic between service A and service B was successful, or not.  As @lei-tang notes, I'm not sure how we would correlate the two legs in waypoint-double to show the request and its response time etc.  I do see in both diagrams that src and dst workloads as noted as pods, is that correct?  I don't think we'd want pods.

**Comment by howardjohn** [comment]
> src and dst workloads as noted as pods, is that correct? I don't think we'd want pods.

This was a bit of a misnomer, it would be the standard app/canonicalservice/workload labels ,not a new 'pod' one.

---

It seems like there is some consensus building around the "waypoint single" option, with the ztunnel metrics also shown there (which also happen to be what ztunnel implements today).

**Comment by linsun** [comment]
Is it correct to think waypoint double (top diagram) provides time spent in a given waypoint while waypoint single (botoom diagram) doesn't provide that info?

Personally i found this is very hard to compare with sidecars because waypoint single or double doesn't report time spent between source and destination.

**Comment by howardjohn** [comment]
> Is it correct to think waypoint double (top diagram) provides time spent in a given waypoint while waypoint single (botoom diagram) doesn't provide that info?


I don't think it provides much meaningful information. The waypoint itself is reporting it during its processing; its hard to accurately observe your own latency.

For example, consider a request takes 10ms of processing

* 0ms: request packet hits waypoint
* 5ms: waypoint done processing it, reports it got it
* 6ms: waypoint about to send upstream request, reports its sending
* 10ms: request finally leaves the waypoint fully

In reality it was 10ms, but we observe 1ms

**Comment by linsun** [comment]
Thank you @howardjohn and @kyessenov - I agree waypoint double doesn't provide meaningful latency within waypoint itself.  And latency within waypoint is out of scope for this WI.

**Comment by lei-tang** [comment]
We drafted a "Plan for OSS ambient mesh beta - telemetry" (https://docs.google.com/document/d/15bW22HxuB9F-TIDy4h_YCBsUb8usbaurx8-DrXcL3_Q/edit). Please take a look.

**Comment by howardjohn** [comment]
I've temporarily lost write access to docs, so will comment here:
* Ztunnel should write some logs, but not envoy logs format which were designed for HTTP. We can provide a far better log.
* Waypoint should support tracing just like Istio Gateway does. 
* -1 to decision to not do Telemetry API. Its probably more work to _not_ support it
* Should remove "Additional stats for waypoint" or make it clear its just providing more context, not proposing we do that

Looks good though, thanks for pulling that together.

**Comment by lei-tang** [comment]
John, thank you for your comments! I agree with your points and have updated the doc accordingly. PTAL.

**Comment by linsun** [comment]
To recap what we discussed recently in ambient meeting, I think we have a plan for ambient beta, which is to go with waypoint single for reporting (bottom of this [diagram](https://github.com/istio/istio/issues/42320#issuecomment-1977262417) and not support Telemetry API.  We need to identify an owner for #50225.
